### PR TITLE
Slide improvements pass

### LINKED ISSUE #186: "Promote" `other_keys` to be printed, a constructor parameter, and more clearly documented
Migrated from #113 and #183.
- `print.epi_df` should print the `other_keys` metadata.
- `other_keys` should be its own parameter to `new_epi_df` and `as_epi_df`
- Unfortunately, I don't think we ever explained what "keys" were in the context of an `epi_df`, and googling "tibble key" is surprisingly unhelpful (currently giving links to lock picks and decoder tools) [so the `{new,as}_epi_df` docs may need explication].  Maybe explaining this could just be rolled into the issue for adding `other_keys` as a parameter, though; it'd be easier to do in its own `@param`.

If this is causing recurring issues in epipredict maybe it should be changed to higher priority.

### LINKED ISSUE #240: Consider allowing list-type output from `epi[x]_slide` inner computations
Currently, a workaround for the length/nrow requirement for the inner computations is to return a single-row data frame with a single list column containing the actual result, then do a `time_value`-grouped slice and maybe a `select` to drop the nongrouping key columns on the slide output.  This could be slightly less work if one could just say `data = list(<real result>)` for the computation rather than `data = tibble(result = <real result>)`.

However, we might want to watch out for users trying to achieve something like the `%>% mutate(tibble(col1=..., col2=...))` behavior, except with lists instead of tibbles.  Perhaps that's enough to dissuade us from implementing this, or maybe there's a way around it; e.g., accepting only unnamed lists, since the names would probably be dropped anyway.

Implementation might require a tweak of `as_list_col`, not just the validation code.

### LINKED ISSUE #446: [enh] promote `other_keys`
Can we move `other_keys` out of the `additional_metadata` List? This seems like, by far, the most common use case we've seen for this list. And it already has its own special handling in the `new_epi_df()` constructor.

### LINKED ISSUE #510: `epi_slide` and `epix_slide` args
I still find before/after confusing. It takes mental math and careful reading of the documentation. It's more general, but not really how our target users are likely to think.

I suggest we bring back `n` for the window size as well as `align = c("left", "center", "right")`. We can keep `before`/`after` as `NULL` and ensure there aren't any conflicts. Then immediately process `n`+`align` into the appropriate `before`+`after` pair.

### PR DESCRIPTION
**Pull request by brookslogan**
### Checklist

Please:

- [x] Make sure this PR is against "dev", not "main" (unless this is a release
      PR).
- [x] Request a review from one of the current main reviewers:
      brookslogan, nmdefries.
- [x] Makes sure to bump the version number in `DESCRIPTION`. Always increment
      the patch version number (the third number), unless you are making a
      release PR from dev to main, in which case increment the minor version
      number (the second number).
- [x] Describe changes made in NEWS.md, making sure breaking changes
      (backwards-incompatible changes to the documented interface) are noted.
      Collect the changes under the next release number (e.g. if you are on
      1.7.2, then write your changes under the 1.8 heading).
- See [DEVELOPMENT.md](DEVELOPMENT.md) for more information on the development
  process.

### Magic GitHub syntax to mark associated Issue(s) as resolved when this is merged into the default branch

- Resolves #240
- Resolves #446 
- Resolves #186 
- Resolves #510 
- Resolves (partially?) #162 
- Addresses some (but not all) naming issues here #146

### CODE REVIEW

**Comment by brookslogan** [comment]
(Random implementation notes: I also investigated an `as_slide_computations()` approach that would give a length>=1 list of functions to apply in the tidyeval case and length-1 lists in the function and formula cases.  Pros of that way include reducing code duplication/complexity/divergence when validating computation outputs, and having more freedom if we want to try to detect and optimize some slide computations by transforming them to calls to (internals of) `epi_slide_opt()` (as dplyr grouped_df's already do or did at some point).  Cons of that way are that we may introduce some computational overhead in the function and formula cases (setting up, storing, munging `results_env`s --- this is assuming the loop over computations has been hoisted above the group and/or ref time values loops to enable other optimizations; if that's not the other case, then there's the overhead of repeatedly executing a loop over 1 function which might actually be relevant if you're doing an `epi_slide()` with an unrecognized simple function) + need for yet another argument (.data / results, not just .x --- either added user-side or via transformations which also have time/debuggability overhead) or code complexity to avoid this.)

**Comment by brookslogan** [comment]
- [ ] doc parm deprecations

File: R/utils.R [hunk 1]
```diff
@@ -283,22 +283,78 @@ as_slide_computation <- function(f, ...) {
...
+        } else {
+          cli_abort("
+            Problem with output of {.code
+            {rlang::expr_deparse(rlang::quo_get_expr(f[[quosure_i]]))}}; it
+            produced a result that was neither NULL, a data.frame, nor a vector
+            without unnamed entries (as determined by the vctrs package).
+          ", class = "epiprocess__invalid_slide_comp_tidyeval_output")
+        }
+      }
+      validate_tibble(new_tibble(as.list(results_env, all.names = TRUE)[results_names]))
```
Comment (brookslogan) [inline R/utils.R:357]: issue?: in rare case that user redefines a col within a computation, we'll request same colname twice here.  Somehow gets de-duped down the line, but probably better to make some comments or fixes here.

**Comment by brookslogan** [comment]
In retrospect, this round of changes isn't as helpful or necessary as I expected.

The main change here seems to be
- Defaulting to something like `names_sep = NULL`.  This requires a second round of changes (incoming) to really work smoothly.  This second pass is probably going to much more impactful.

There are some potential minor conveniences: 
- Multiple slide computations
- `:=` support
- Hiding a bit more easily from `epi_slide` output size checks (`as_list_col = TRUE` wasn't enough, because size checks happened before list conversion)..... but in cases where you want to do this, you still want to slice away duplicate results, still quite awkward.
  - Hiding from `epix_slide()` naming is maybe the same level of convenience as before, though it feels a bit more natural to me now to wrap in `list()` rather than `as_list_col = TRUE`.  Another thing that's actually addressed by round 2 and doesn't really need this round 1 except for the convenience of not adding column prefixes by default.
- Maybe a little more parallelism between outputting data frames out of tidyeval and nontidyeval approaches.  But they are still their own distinct things.
- Better parallelism with dplyr syntax/expectations, at least for advanced users.
- Potentially setting up better for the next item below.

What I should probably add on to this pass or another PR:
- Support for slide computations directly outputting the result of `arx_forecaster()` ("scalar" S3 things).

And what neither of these rounds are going to address:
- Lack of `epi_slide_{summarize,reframe}` functions.  This is what we might want for some geoaggregation and version-unfaithful forecasting applications..... but there are probably better ways requiring less work.

**Comment by brookslogan** [comment]
- [ ] todo: see if can rebase to have a `names_sep = NULL` default commit

**Comment by dshemetov** [comment]
We should probably bump the version to 0.9.0, since that's our way of signaling breaking changes. Should probably also tag the version prior to merging this in, to make it easy to pin.

**Comment by dsweber2** [comment]
I was talking with Dmitry about this, and I would request any functions that are defined in-line inside other functions be moved outside and have their dependencies included as explicit arguments, instead of capturing their value from context. They make it hard to follow the thread of what a function is doing.

**Comment by brookslogan** [comment]
I accidentally pushed some changes that introduced some more dot prefixes to try to prevent some confusion and edge cases.  But I think this also fixed some outdated error messages, so maybe it's a good thing.  (There are also some orthogonal but also renaming-related fixes mixed in as well.)

- [ ] todo: fix roxygen/pkgdown config to solve pkgdown issue

**Comment by dshemetov** [comment]
- [ ] consider a rename of epi_slide / epix_slide (but let's try not to get stuck here)

File: R/utils.R (hunk 1 above)
Comment (brookslogan) [reply R/utils.R]: should be done

**Review by dshemetov — APPROVED** [review]
Good to go. Will merge to main after this. epipredict PR will follow soon after.
